pkg/controllers/workgenerator/controller.go

```diff
@@ -224,52 +231,98 @@ func (r *Reconciler) listAllWorksAssociated(ctx context.Context, resourceBinding
 
 	// Gather all the resource resourceSnapshots
 	resourceSnapshots, err := r.fetchAllResourceSnapshots(ctx, resourceBinding)
 	if err != nil {
-		// TODO(RZ): handle errResourceNotFullyCreated error
+		// TODO(RZ): handle errResourceNotFullyCreated error so we don't need to wait for all the snapshots to be created
 		return false, err
 	}
 
-	// create/update the corresponding work for each snapshot
-	activeWork := make(map[string]bool, len(resourceSnapshots))
-	for _, snapshot := range resourceSnapshots {
-		// TODO(RZ): issue those requests in parallel to speed up the process
-		updated := false
-		workName, err := getWorkNameFromSnapshotName(snapshot)
+	// issue all the create/update requests for the corresponding works for each snapshot in parallel
+	activeWork := make(map[string]*fleetv1beta1.Work, len(resourceSnapshots))
+	errs, cctx := errgroup.WithContext(ctx)
+	// generate work objects for each resource snapshot
+	for i := range resourceSnapshots {
+		snapshot := resourceSnapshots[i]
+		var newWork []*fleetv1beta1.Work
+		workNamePrefix, err := getWorkNamePrefixFromSnapshotName(snapshot)
 		if err != nil {
 			klog.ErrorS(err, "Encountered a mal-formatted resource snapshot", "resourceSnapshot", klog.KObj(snapshot))
 			return false, err
 		}
-		activeWork[workName] = true
-		if updated, err = r.upsertWork(ctx, works[workName], workName, snapshot, resourceBinding); err != nil {
-			return false, err
+		var simpleManifests []fleetv1beta1.Manifest
+		for _, selectedResource := range snapshot.Spec.SelectedResources {
+			// we need to special treat configMap with envelopeConfigMapAnnotation annotation,
+			// so we need to check the GVK and annotation of the selected resource
+			var uResource unstructured.Unstructured
+			if err := uResource.UnmarshalJSON(selectedResource.Raw); err != nil {
+				klog.ErrorS(err, "work has invalid content", "snapshot", klog.KObj(snapshot), "selectedResource", selectedResource.Raw)
+				return false, controller.NewUnexpectedBehaviorError(err)
+			}
+			if uResource.GetObjectKind().GroupVersionKind() == utils.ConfigMapGVK &&
+				len(uResource.GetAnnotations()[fleetv1beta1.EnvelopeConfigMapAnnotation]) != 0 {
+				// get a work object for the enveloped configMap
+				work, err := r.getConfigMapEnvelopWorkObj(ctx, workNamePrefix, resourceBinding, snapshot, &uResource)
+				if err != nil {
+					return false, err
+				}
+				newWork = append(newWork, work)
+			} else {
+				simpleManifests = append(simpleManifests, fleetv1beta1.Manifest(selectedResource))
+			}
 		}
-		if updated {
-			updateAny = true
+		if len(simpleManifests) != 0 {
+			// generate a work object for the manifests if there are still any non enveloped resources
+			newWork = append(newWork, generateSnapshotWorkObj(workNamePrefix, resourceBinding, snapshot, simpleManifests))
+		} else {
+			klog.V(2).InfoS("Skip generating work for the snapshot since there is no none-enveloped resource in the snapshot", "snapshot", klog.KObj(snapshot))
```

Comment:
Hi Ryan! A nit here: this message seems to be only suggesting that there're only unwrapped resources; the skip part reads a little bit confusing to me personally.